lib/core/src/Cardano/Wallet/Primitive/Scripts.hs

```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of wallet's verification key discovery
+-- in a script.
+
+module Cardano.Wallet.Primitive.Scripts
+    ( isShared
+    , retrieveAllVerKeyHashes
+    ) where
+
+import Prelude
+
+import Cardano.Address.Script
+    ( KeyHash (..), Script (..), ScriptHash (..), toScriptHash )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationType (..)
+    , Index (..)
+    , Role (..)
+    , SoftDerivation (..)
+    )
+import Cardano.Wallet.Primitive.AddressDerivation.Shelley
+    ( ShelleyKey (..) )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( SeqState (..)
+    , VerificationKeyPool (..)
+    , defaultAddressPoolGap
+    , getAddressPoolGap
+    , mkUnboundedAddressPoolGap
+    , mkVerificationKeyPool
+    )
+import Cardano.Wallet.Primitive.Types.Address
+    ( AddressState (..) )
+import Control.Monad
+    ( foldM )
+import Data.Functor.Identity
+    ( Identity (..) )
+
+import qualified Data.List as L
+import qualified Data.Map.Strict as Map
+
+instance Ord ScriptHash where
+    compare (ScriptHash sh1) (ScriptHash sh2) = compare sh1 sh2
+
+isShared
+    :: (k ~ ShelleyKey, SoftDerivation k)
+    => Script
+    -> SeqState n k
+    -> ([k 'ScriptK XPub], SeqState n k)
+isShared script (SeqState !s1 !s2 !ixs !rpk !prefix !scripts !s3) =
+    let verKeysInScript = retrieveAllVerKeyHashes script
+        (VerificationKeyPool accXPub currentGap verKeyMap) = s3
+        projectKey (ShelleyKey k) = ShelleyKey k
+        toVerKey = deriveAddressPublicKey accXPub MultisigScript
+        projectIndex = toEnum . fromInteger . toInteger . getIndex @'Soft
+        ourVerKeysInScript =
+            map (\(_,(ix,_)) -> toVerKey (projectIndex ix) ) $
+            filter (\(keyH,_) -> keyH `elem` verKeysInScript)
+            (Map.toList verKeyMap)
+        scriptXPubs = L.nub $ map projectKey ourVerKeysInScript
+        updateAddressState k current =
+            if k `elem` verKeysInScript then
+                Used
+            else
+                current
+        markedVerKeyMap =
+            Map.fromList $
+            map (\(keyH,(ix,isUsed)) -> (keyH, (ix, updateAddressState keyH isUsed)) )
+            (Map.toList verKeyMap)
+        -- if all verification keys are used (after discovering) we are extending multisigPool
```

Comment:
Is that the case though? We want to extend the pool as soon as a _single_ key is found and break the invariant: there should always be `gap` consecutive unused items.

Reply:
how gap behaves and influence pool size was completely refactored here https://github.com/input-output-hk/cardano-wallet/pull/2308/commits/1a862ab8699cc2d46f84258c547048d085d7dcf5